### PR TITLE
[Builder] Support pushing images to ECR

### PR DESCRIPTION
**Pull request by AlonMaor14**
Changes to kaniko pod to support ECR:
- Volume aws secret generated from .aws/credentials file to communicate with ECR.
- Use init container with aws cli to create the ECR repository if it does not exist.
- If aws secret was not given - assume instance role

### CODE REVIEW

**Review by quaark — APPROVED** [review]
LGTM

**Review by Hedingber — CHANGES_REQUESTED** [review]

**Review by Tankilevitch — CHANGES_REQUESTED** [review]

File: mlrun/config.py [hunk 1]
```diff
@@ -277,6 +277,11 @@
             "mlrun_version_specifier": "",
             "kaniko_image": "gcr.io/kaniko-project/executor:v1.8.0",  # kaniko builder image
             "kaniko_init_container_image": "alpine:3.13.1",
+            # image for kaniko init container when docker registry is ECR
+            "kaniko_aws_cli_image": "amazon/aws-cli:2.7.10",
+            # auth secret to attach to kaniko pod for off cluster registries
+            # for example: aws-secret in https://github.com/GoogleContainerTools/kaniko#pushing-to-amazon-ecr
+            "kaniko_registry_provider_auth_secret": "",
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/config.py:284]: Structure for the config should be
```
"function": {
        "spec":{
                "image_pull_secret":{
                    "default": "secret-name"
         },
}
```

File: mlrun/builder.py [hunk 2]
```diff
@@ -171,15 +173,62 @@ def make_kaniko_pod(
...
+                "AWS_SHARED_CREDENTIALS_FILE": "/tmp/credentials",
+            },
+            name="create-repo",
+        )
+
+        # set the kaniko container AWS credentials location to the mount's path
+        kpod.env.append(
+            client.V1EnvVar(
+                name="AWS_SHARED_CREDENTIALS_FILE", value="/tmp/credentials"
+            )
```
[the remark this answers is not in the document]
Comment (AlonMaor14) [reply mlrun/builder.py:228]: the 1st one is for the init container
the 2nd one is for the kaniko container

File: mlrun/runtimes/pod.py [hunk 3]
```diff
@@ -147,7 +147,9 @@ def __init__(
         self.replicas = replicas
         self.image_pull_policy = image_pull_policy
         self.service_account = service_account
-        self.image_pull_secret = image_pull_secret
+        self.image_pull_secret = (
+            image_pull_secret or mlrun.mlconf.httpdb.builder.docker_registry_secret
+        )
```
[the remark this answers is not in the document]
Comment (AlonMaor14) [reply mlrun/runtimes/pod.py:152]: since we changed the default spec I'm assuming this is no longer relevant

**Review by Hedingber — APPROVED** [review]

File: tests/runtimes/test_run.py [hunk 4]
```diff
@@ -66,6 +67,7 @@ def test_new_function_args_without_command():
     expected_runtime["spec"][
         "preemption_mode"
     ] = mlrun.mlconf.function_defaults.preemption_mode
+    expected_runtime["spec"]["image_pull_secret"] = ""
```
[the remark this answers is not in the document]
Comment (Tankilevitch) [reply tests/runtimes/test_run.py:70]: Instead add test that checks the auto enrichment of the default and also the overriding of the default.

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
Looks good, added minor comments.

**Review by Tankilevitch — APPROVED** [review]

**Pull request merged**